node/pkg/driver/node.go

```diff
@@ -345,47 +341,107 @@ func (d *NodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	mpathDevice := filepath.Join(device_connectivity.DevPath, infoMap["mpathDevice"])
 	logger.Debugf("Got info from stageInfo file. device : {%v}", mpathDevice)
 
-	logger.Debugf("Check if targetPath {%s} exist in mount list", targetPath)
-	mountList, err := d.mounter.List()
-	for _, mount := range mountList {
-		logger.Tracef("Check if mount path({%v}) [with device({%v})] is equel to targetPath {%s}", mount.Path, mount.Device, targetPath)
-		if mount.Path == targetPathWithHostPrefix {
-			if mount.Device == mpathDevice {
-				logger.Warningf("Idempotent case : targetPath already mounted (%s), so no need to mount again. Finish NodePublishVolume.", targetPath)
-				return &csi.NodePublishVolumeResponse{}, nil
-			} else {
-				return nil, status.Errorf(codes.AlreadyExists, "Mount point is already mounted to but with different multipath device (%s), while the expected device is %s ", mount.Device, mpathDevice)
-			}
+	// if the device is not mounted then we are mounting it.
+	volumeCap := req.GetVolumeCapability()
+	isFSVolume := true
+	switch volumeCap.GetAccessType().(type) {
+	case *csi.VolumeCapability_Block:
+		isFSVolume = false
+	}
+	isTargetPathExists := d.NodeUtils.IsPathExists(targetPathWithHostPrefix)
+	if isTargetPathExists {
+		// check if already mounted
+		isMounted, err := d.isTargetMounted(targetPath, isFSVolume)
```

Comment:
A common way is use mounter.IsLikelyNotMountPoint(), see the code here:
https://github.com/kubernetes/utils/blob/master/mount/mount_linux.go#L222

Reply:
This is how it was done before my change. Also we'll add additional validation here which is not checked by IsLikelyNotMountPoint

Reply:
Lue I don't intend to refactor most of the code as part of this ticket.

Reply:
If this performance issue was there before this change, then I think we need a ticket for it

Reply:
ok, please open the ticket and attach the number here.

Reply:
please don't mark this conversation as resolved, it is not resolved yet.